File: tests/backend/unit/rendering_test.rb
# frozen_string_literal: true

require_relative '../../test_helper'
require 'services/rendering'

# TF-101 to TF-134 — the rendering pipeline (Requirements chapter 8).
#
# The cases are not written out here. They live in tests/vectors/rendering.json
# and are read from that file, because the Vitest suite in AP-11 reads the very
# same file. Two copies of the expected values would drift apart, and the whole
# point of these vectors is that both implementations answer to one source.
#
# A failure in this suite means one of three things (test concept 5.6):
#   both sides wrong the same way  -> specification misread, fix both
#   only Ruby wrong                -> fix here
#   the two sides differ           -> risk R-01, delivery blocked
class RenderingTest < PromptAtelier::TestCase
  R = PromptAtelier::Rendering

  VECTOR_FILE = File.join(CODE_ROOT, 'tests', 'vectors', 'rendering.json')
  DATA        = JSON.parse(File.read(VECTOR_FILE, encoding: 'UTF-8')).freeze
  VECTORS     = DATA['vectors'].freeze

  # --- the vectors themselves ---------------------------------------------

  # One test method per vector, generated at load time. Generated rather than
  # looped inside a single test so that a failure names the vector — "TF-119
  # failed" is a place to look, "the vector test failed" is not.
  VECTORS.each do |vector|
    id   = vector['id']
    name = id.downcase.tr('-', '_')

    define_method("test_#{name}_#{vector['title'].downcase.gsub(/[^a-z0-9]+/, '_')}") do
      result = render_vector(vector)

      assert_equal vector['expected'], result.text,
                   "#{id} (#{vector['title']}) does not match character for character"

      if vector.key?('unknown_keys')
        assert_equal vector['unknown_keys'], result.unknown_keys,
                     "#{id}: unexpected set of unknown keys"
      end

      if vector.key?('missing_required')
        assert_equal vector['missing_required'], result.missing_required,
                     "#{id}: unexpected set of unbound required variables"
      end
    end
  end

  # --- the vector file as such ---------------------------------------------

  # A-12 promises 34 vectors. If one is dropped the suite above would simply
  # run one test fewer and stay green — silently.
  def test_the_file_holds_exactly_the_34_documented_vectors
    ids = VECTORS.map { |vector| vector['id'] }

    assert_equal 34, ids.size
    assert_equal (101..134).map { |number| "TF-#{number}" }, ids,
                 'TF-101 to TF-134, in order and without gaps'
    assert_equal ids.uniq, ids, 'no duplicates'
  end

  def test_every_vector_carries_a_body_and_an_expectation
    VECTORS.each do |vector|
      refute_nil vector['body'],     "#{vector['id']}: body missing"
      refute_nil vector['expected'], "#{vector['id']}: expected missing"
      refute_empty vector['title'].to_s, "#{vector['id']}: title missing"
    end
  end

  def test_every_named_keyword_is_defined
    VECTORS.each do |vector|
      known = DATA['keywords'].merge(vector['extra_keywords'] || {})
      (vector['keywords'] || []).each do |name|
        refute_nil known[name], "#{vector['id']}: keyword #{name.inspect} is not defined"
      end
    end
  end

  # The fixture set from test concept 4.4, unchanged. If someone edits a
  # keyword text here, several expectations further down become wrong in a way
  # that is hard to trace back.
  def test_the_shared_keywords_match_the_test_concept
    expected = {
      'rolle'   => ['prepend', 10, 'Du bist ein erfahrener Fachautor.'],
      'formal'  => ['append',  10, 'Schreibe in einem sachlichen, förmlichen Ton.'],
      'kurz'    => ['append',  20, 'Fasse dich kurz, höchstens 150 Wörter.'],
      'anhang'  => ['append',  20, 'Nenne am Ende zwei Quellen.'],
      'kontext' => ['prepend', 20, 'Der Text erscheint in einem Fachblog.']
    }

    assert_equal expected.keys.sort, DATA['keywords'].keys.sort
    expected.each do |name, (position, order, text)|
      keyword = DATA['keywords'][name]
      assert_equal position, keyword['position'], name
      assert_equal order,    keyword['sort_order'], name
      assert_equal text,     keyword['text'], name
    end
  end

  # --- properties the vectors do not cover ---------------------------------

  # NFA-14: the pipeline must be usable without the web layer or the database.
  # It is loaded here on its own, so this passes only as long as it stays that
  # way.
  def test_the_pipeline_needs_neither_sinatra_nor_a_database
    source = File.read(File.join(CODE_ROOT, 'backend', 'services', 'rendering.rb'))

    refute_match(/require.*sinatra/, source)
    refute_match(/require.*sequel/i, source)
    refute_match(/require_relative/, source, 'no dependency on other services either')
  end

  def test_rendering_the_same_input_twice_gives_the_same_result
    VECTORS.each do |vector|
      first  = render_vector(vector).text
      second = render_vector(vector).text
      assert_equal first, second, "#{vector['id']} is not deterministic"
    end
  end

  # --- key recognition (8.2), beyond the vectors ---------------------------

  # TF-404 as an edge case: a key with a space is not a variable at all — it
  # produces no warning either, because nothing was recognised.
  def test_tf404_a_key_with_a_space_is_not_a_variable_and_warns_about_nothing
    result = R.render(body: 'Nutze {{Mein Thema}} hier.')

    assert_equal 'Nutze {{Mein Thema}} hier.', result.text
    assert_empty result.unknown_keys
  end

  # 8.2 asks for the lower-casing on **both** sides of the comparison: in the
  # text and in the record. The text is covered by the vectors, the record was
  # covered on neither side — both versions lower-case, neither was ever held
  # to it. Whoever strikes one of the two lines notices it nowhere; that is the
  # construction R-01 warns about, only doubled.
  def test_the_key_of_the_record_is_compared_without_regard_to_case
    result = R.render(body: '{{thema}}', variables: [{ key: 'THEMA', value: 'Kaffee' }])

    assert_equal 'Kaffee', result.text
    assert_empty result.unknown_keys
  end

  def test_keys_are_bounded_to_forty_characters
    fits    = 'a' * 40
    too_big = 'a' * 41

    assert_equal 'X', R.render(body: "{{#{fits}}}",
                               variables: [{ key: fits, value: 'X' }]).text
    # Not a variable, so it stays put and raises no warning.
    result = R.render(body: "{{#{too_big}}}", variables: [{ key: too_big, value: 'X' }])
    assert_equal "{{#{too_big}}}", result.text
    assert_empty result.unknown_keys
  end

  def test_a_hyphen_or_a_dot_does_not_belong_in_a_key
    ['{{mein-thema}}', '{{mein.thema}}', '{{ }}', '{{}}'].each do |text|
      assert_equal text, R.render(body: text).text, "#{text} must stay as it is"
    end
  end

  def test_underscores_and_digits_are_allowed_after_the_first_letter
    result = R.render(body: '{{thema_2}}', variables: [{ key: 'thema_2', value: 'Kaffee' }])

    assert_equal 'Kaffee', result.text
  end

  # The same unknown key twice is one warning, not two — and in the order of
  # first occurrence. Neither implementation pinned this until a mutation
  # probe on the JavaScript side removed the check and every vector stayed
  # green.
  def test_an_unknown_key_is_reported_once_in_order_of_first_occurrence
    result = R.render(body: '{{b}} {{a}} {{b}} {{a}}')

    assert_equal %w[b a], result.unknown_keys
  end

  # --- values --------------------------------------------------------------

  def test_a_missing_binding_falls_back_to_the_default_then_to_empty_text
    with_default = R.render(body: '{{a}}.', variables: [{ key: 'a', default_value: 'D' }])
    without      = R.render(body: '{{a}}.', variables: [{ key: 'a' }])

    assert_equal 'D.', with_default.text
    assert_equal '.',  without.text
  end

  # A value of "0" or "false" is a value. Falling back to the default there
  # would be a classic truthiness bug.
  def test_the_string_zero_counts_as_a_binding
    result = R.render(body: '{{a}}', variables: [{ key: 'a', value: '0', default_value: 'D' }])

    assert_equal '0', result.text
  end

  # --- required variables (8.3) --------------------------------------------

  def test_a_required_variable_with_a_default_is_not_reported_as_missing
    result = R.render(body: '{{a}}',
                      variables: [{ key: 'a', required: true, default_value: 'D' }])

    assert_equal 'D', result.text
    assert_empty result.missing_required
    assert result.complete?
  end

  def test_the_preview_is_produced_even_when_a_required_value_is_missing
    result = R.render(body: 'Schreibe über {{thema}}.',
                      variables: [{ key: 'thema', required: true }])

    assert_equal 'Schreibe über .', result.text, '8.3: the preview is still rendered'
    assert_equal ['thema'], result.missing_required
    refute result.complete?, 'copying is what gets blocked, not the rendering'
  end

  # --- keywords ------------------------------------------------------------

  def test_an_empty_keyword_text_adds_no_separator
    result = R.render(body: 'Text.',
                      keywords: [{ name: 'leer', text: '', position: 'append', sort_order: 10 }])

    assert_equal 'Text.', result.text
  end

  def test_keyword_order_does_not_depend_on_the_order_they_arrive_in
    keywords = [
      { name: 'b', text: 'Zweitens.', position: 'append', sort_order: 20 },
      { name: 'a', text: 'Erstens.',  position: 'append', sort_order: 10 }
    ]

    forwards  = R.render(body: 'Text.', keywords: keywords).text
    backwards = R.render(body: 'Text.', keywords: keywords.reverse).text

    assert_equal "Text.\n\nErstens.\n\nZweitens.", forwards
    assert_equal forwards, backwards
  end

  private

  def render_vector(vector)
    keywords = (vector['keywords'] || []).map do |name|
      definition = (vector['extra_keywords'] || {}).fetch(name) { DATA['keywords'].fetch(name) }
      definition.merge('name' => name)
    end

    PromptAtelier::Rendering.render(
      body: vector['body'],
      variables: vector['variables'] || [],
      keywords: keywords
    )
  end
end

File: tests/frontend/rendering.test.js
// @vitest-environment node
//
// No DOM is involved: the pipeline is plain data in, plain data out (NFA-14),
// and it stays that way only as long as nothing here needs a browser.

import { describe, it, expect } from 'vitest'
import { readFileSync } from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { render, variableKeys, measure } from '../../frontend/src/util/rendering.js'

// TF-101 to TF-134 — the same 34 vectors the Minitest suite runs.
//
// This is the proof for A-12 and the safeguard against R-01. Both sides read
// **this file**, not a copy: two lists of expected values would drift apart,
// and the whole point is that the two implementations answer to one source.
//
// A failure means one of three things (test concept 5.6):
//   both sides wrong the same way  -> specification misread, fix both
//   only JavaScript wrong          -> fix here
//   the two sides differ           -> risk R-01, delivery blocked

const HERE = path.dirname(fileURLToPath(import.meta.url))
const VECTOR_FILE = path.resolve(HERE, '..', 'vectors', 'rendering.json')
const DATA = JSON.parse(readFileSync(VECTOR_FILE, 'utf8'))
const VECTORS = DATA.vectors

function renderVector (vector) {
  const keywords = (vector.keywords ?? []).map((name) => {
    const definition = (vector.extra_keywords ?? {})[name] ?? DATA.keywords[name]
    return { ...definition, name }
  })

  return render({
    body: vector.body,
    variables: vector.variables ?? [],
    keywords
  })
}

describe('Rendering vectors', () => {
  // One case per vector rather than one loop, so a failure names the vector:
  // "TF-119 failed" is a place to look, "the vector test failed" is not.
  for (const vector of VECTORS) {
    it(`${vector.id} — ${vector.title}`, () => {
      const result = renderVector(vector)

      expect(result.text).toBe(vector.expected)

      if ('unknown_keys' in vector) expect(result.unknownKeys).toEqual(vector.unknown_keys)
      if ('missing_required' in vector) expect(result.missingRequired).toEqual(vector.missing_required)
    })
  }
})

describe('The vector file', () => {
  // A-12 promises 34 vectors. If one were dropped the suite above would
  // simply run one case fewer and stay green — silently.
  it('carries exactly the 34 documented vectors', () => {
    const ids = VECTORS.map((vector) => vector.id)

    expect(ids).toHaveLength(34)
    expect(ids).toEqual(Array.from({ length: 34 }, (unused, index) => `TF-${101 + index}`))
  })

  // The file is the shared source. If this side ever read a copy of its own,
  // the safeguard against R-01 would be gone and nothing would say so.
  it('reads the same file as the Ruby side', () => {
    expect(VECTOR_FILE).toBe(path.resolve(HERE, '..', 'vectors', 'rendering.json'))
    expect(readFileSync(VECTOR_FILE, 'utf8')).toContain('Diese Datei wird von Minitest UND von Vitest gelesen')
  })
})

// The properties the vectors do not pin down — the same set the Minitest
// suite checks, case for case.
//
// They belong on both sides for the same reason the vectors do: a rule that
// only one implementation is held to is a rule the other may break. A
// mutation probe found three of these missing here while Ruby had them, which
// is exactly the asymmetry R-01 warns about.
describe('Key detection (8.2)', () => {
  it('takes a key with a space for no key and warns about nothing', () => {
    const result = render({ body: 'Nutze {{Mein Thema}} hier.' })

    expect(result.text).toBe('Nutze {{Mein Thema}} hier.')
    expect(result.unknownKeys).toEqual([])
  })

  it('limits a key to forty characters', () => {
    const fits = 'a'.repeat(40)
    const tooBig = 'a'.repeat(41)

    expect(render({ body: `{{${fits}}}`, variables: [{ key: fits, value: 'X' }] }).text).toBe('X')

    const result = render({ body: `{{${tooBig}}}`, variables: [{ key: tooBig, value: 'X' }] })
    expect(result.text).toBe(`{{${tooBig}}}`)
    expect(result.unknownKeys).toEqual([])
  })

  it('does not let hyphen, dot and emptiness count as a key', () => {
    for (const text of ['{{mein-thema}}', '{{mein.thema}}', '{{ }}', '{{}}']) {
      expect(render({ body: text }).text).toBe(text)
    }
  })

  it('allows underscores and digits after the first letter', () => {
    expect(render({ body: '{{thema_2}}', variables: [{ key: 'thema_2', value: 'Kaffee' }] }).text)
      .toBe('Kaffee')
  })

  // 8.2 asks for the lower-casing on **both** sides of the comparison: in the
  // text and in the record. The text is covered by the vectors, the record was
  // covered on neither side — both versions lower-case, neither was ever held
  // to it. Exactly the construction R-01 warns about, only doubled: whoever
  // strikes one of the two lines notices it nowhere.
  it('compares the key of the record without regard to case as well', () => {
    const result = render({ body: '{{thema}}', variables: [{ key: 'THEMA', value: 'Kaffee' }] })

    expect(result.text).toBe('Kaffee')
    expect(result.unknownKeys).toEqual([])
  })

  // The same unknown key twice is one warning, not two — and in the order of
  // first occurrence. Neither side pinned this until a mutation probe removed
  // the check and every vector stayed green.
  it('names an unknown key once, in the order of appearance', () => {
    const result = render({ body: '{{b}} {{a}} {{b}} {{a}}' })

    expect(result.unknownKeys).toEqual(['b', 'a'])
  })
})

describe('Values', () => {
  it('falls back to the default value and then to empty text', () => {
    expect(render({ body: '{{a}}.', variables: [{ key: 'a', default_value: 'D' }] }).text).toBe('D.')
    expect(render({ body: '{{a}}.', variables: [{ key: 'a' }] }).text).toBe('.')
  })

  // A value of "0" is a value. Falling back to the default there would be the
  // classic truthiness mistake — and in JavaScript it is one line away.
  it('takes zero as a value', () => {
    expect(render({ body: '{{a}}', variables: [{ key: 'a', value: '0', default_value: 'D' }] }).text)
      .toBe('0')
    expect(render({ body: '{{a}}', variables: [{ key: 'a', value: 0, default_value: 'D' }] }).text)
      .toBe('0')
  })
})

describe('Required variables (8.3)', () => {
  it('does not report a required variable with a default as unfilled', () => {
    const result = render({ body: '{{a}}', variables: [{ key: 'a', required: true, default_value: 'D' }] })

    expect(result.text).toBe('D')
    expect(result.missingRequired).toEqual([])
    expect(result.complete).toBe(true)
  })

  it('produces the preview even without the required value', () => {
    const result = render({
      body: 'Schreibe über {{thema}}.',
      variables: [{ key: 'thema', required: true }]
    })

    expect(result.text).toBe('Schreibe über .')
    expect(result.missingRequired).toEqual(['thema'])
    // Copying is what gets blocked, not the rendering.
    expect(result.complete).toBe(false)
  })
})

describe('Keywords', () => {
  it('inserts no separator for an empty keyword text', () => {
    const result = render({
      body: 'Text.',
      keywords: [{ name: 'leer', text: '', position: 'append', sort_order: 10 }]
    })

    expect(result.text).toBe('Text.')
  })

  it('does not depend on the order in which the keywords arrive', () => {
    const keywords = [
      { name: 'b', text: 'Zweitens.', position: 'append', sort_order: 20 },
      { name: 'a', text: 'Erstens.', position: 'append', sort_order: 10 }
    ]

    const forwards = render({ body: 'Text.', keywords }).text
    const backwards = render({ body: 'Text.', keywords: [...keywords].reverse() }).text

    expect(forwards).toBe('Text.\n\nErstens.\n\nZweitens.')
    expect(backwards).toBe(forwards)
  })
})

describe('Properties beyond the vectors', () => {
  it('yields the same result for the same input', () => {
    for (const vector of VECTORS) {
      expect(renderVector(vector).text).toBe(renderVector(vector).text)
    }
  })

  // The order of the two rules in step 4 that the vectors pin down for the
  // Ruby side. Repeated here because a JavaScript `$` without the m flag
  // matches only the end of the whole text — the single most likely way for
  // the two sides to part company.
  it('removes whitespace at the end of **every** line, not only the last', () => {
    const result = render({ body: 'Zeile A   \nZeile B\t\nZeile C  ' })

    expect(result.text).toBe('Zeile A\nZeile B\nZeile C')
  })

  it('detects the keys of a text in the order of appearance', () => {
    expect(variableKeys('{{Thema}} und {{zielgruppe}} und nochmal {{THEMA}}'))
      .toEqual(['thema', 'zielgruppe'])
    expect(variableKeys('\\{{kein}} Schluessel')).toEqual([])
  })

  // 11.4: characters and words below the preview, for models with a limit.
  it('counts the characters and words of the result', () => {
    expect(measure('Fasse dich kurz.')).toEqual({ characters: 16, words: 3 })
    expect(measure('  mehrere   Leerraeume\n\nund Zeilen ')).toEqual({ characters: 35, words: 4 })
    expect(measure('')).toEqual({ characters: 0, words: 0 })
    // Counted in characters, not in UTF-16 units: an emoji is one character
    // to the reader and two to `String#length`.
    expect(measure('Grüße 😀').characters).toBe(7)
  })
})
